src/main/java/com/avail/optimizer/L2Optimizer.java

```diff
@@ -465,67 +645,170 @@ void postponeConditionallyUsedValues ()
 	 *
 	 * @param instruction
 	 *        The instruction to analyze.
-	 * @param registersConsumedLaterInSameBlock
-	 *        The set of registers which are consumed by instructions after the
-	 *        given one within the same block.  If the given instruction
-	 *        produces an output consumed by a later instruction, the given
-	 *        instruction cannot be moved forward out of its basic block.
+	 * @param registersConsumedLaterInBlock
+	 *        The set of {@link L2Register}s which are consumed by instructions
+	 *        after the given one within the same {@link L2BasicBlock}.  If the
+	 *        given instruction produces an output consumed by a later
+	 *        instruction, the given instruction cannot be moved forward out of
+	 *        its basic block.
+	 * @param semanticValuesConsumedLaterInBlock
+	 *        The set of {@link L2SemanticValue}s which are consumed by
+	 *        instructions after the given one within the same block.  If the
+	 *        given instruction produces an output consumed by a later
+	 *        instruction, the given instruction cannot be moved forward out of
+	 *        its basic block.
+	 * @param readMaskLaterInBlock
+	 *        A mask of {@link HiddenVariable} dependencies that instructions
+	 *        later in the block will read.  Writes to these hidden variables
+	 *        must not be moved past those reads, or the reads won't be able to
+	 *        access the values.
+	 * @param writeMaskLaterInBlock
+	 *        A mask of {@link HiddenVariable} dependencies that instructions
+	 *        later in the block will write.  Reads of these hidden variables
+	 *        must not be moved past those writes, or the reads won't be able to
+	 *        access the values.
+	 * @param candidateTargetEdges
+	 *        The edges that this instruction might be moved through.  These are
+	 *        all outbound edges in instructions that occur later in the current
+	 *        basic block.
 	 * @return The successor {@link L2PcOperand}s through which the instruction
 	 *         can be moved, or {@code null} if the instruction should not move.
 	 */
-	private static @Nullable List<L2PcOperand> successorEdgesToMoveThrough (
+	private static @Nullable Set<L2PcOperand> successorEdgesToMoveThrough (
 		final L2Instruction instruction,
-		final Set<L2Register> registersConsumedLaterInSameBlock)
+		final Set<L2Register> registersConsumedLaterInBlock,
+		final Set<L2SemanticValue> semanticValuesConsumedLaterInBlock,
+		final int readMaskLaterInBlock,
+		final int writeMaskLaterInBlock,
+		final Set<L2PcOperand> candidateTargetEdges)
 	{
-		if (instruction.hasSideEffect()
-			|| instruction.altersControlFlow()
+		if (instruction.altersControlFlow()
 			|| instruction.operation().isPhi()
-			|| instruction.operation().isEntryPoint(instruction))
+			|| instruction.isEntryPoint())
+		{
+			return null;
+		}
+		// TODO Until we've eliminated this flag...
```

Comment:
Does this comment mean that you want `hasSideEffect` to go away eventually?

Reply:
Removed comment.  I believe I was referring to a future world in which instructions and primitives declare the hidden variables that they read or write – and that future is now!